### PR TITLE
[BEAM-2836] Run the "safe" futurize stage, and then fix the parts which parts which weren't "safe"

### PR DESCRIPTION
**Pull request by holdenk**
Run the safe futurize stage, this only broke a few tests (yay), it did introduce a number of long-line issues with lambdas mostly fixed and refactoring them into regular functions.

The two main things that had to be fixed after stage 1 futurize is expected error message in type inference tests and type inference on iterable/tuple access.

To keep people from inadvertently breaking things until we have Python 3 support as part of Jenkins this also updates run_pylint to verify that there are no changes suggested by futurize stage 1.

### CODE REVIEW

**Comment by holdenk** [comment]
Jenkins retest this please.

**Comment by cclauss** [comment]
Very impressive changes!  I am learning just by watching your progress.

**Comment by holdenk** [comment]
Thanks @cclauss , if you ever want to work on a PR together in BEAM or Spark I'd be happy to :)

**Comment by holdenk** [comment]
Ok this should be nice and condensed now. cc @robertwb if you've got the time.

**Comment by holdenk** [comment]
+cc @aaltay

File: sdks/python/apache_beam/examples/complete/game/leader_board.py [hunk 1]
```diff
@@ -331,10 +331,13 @@ def run(argv=None):
...
     }
+
+    def format_user_score_sums(user_score):
+      return {'user': user_score[0], 'total_score': user_score[1]}
+
     (events  # pylint: disable=expression-not-assigned
      | 'CalculateUserScores' >> CalculateUserScores(args.allowed_lateness)
-     | 'FormatUserScoreSums' >> beam.Map(
-         lambda (user, score): {'user': user, 'total_score': score})
+     | 'FormatUserScoreSums' >> beam.Map(format_user_score_sums)
```
Comment (aaltay) [inline sdks/python/apache_beam/examples/complete/game/leader_board.py]: Why is this change needed? (Similarly for other lambda function changes.)

File: sdks/python/apache_beam/examples/complete/juliaset/juliaset/juliaset.py [hunk 2]
```diff
@@ -107,11 +107,14 @@ def run(argv=None):  # pylint: disable=missing-docstring
     # Group each coordinate triplet by its x value, then write the coordinates
     # to the output file with an x-coordinate grouping per line.
     # pylint: disable=expression-not-assigned
```
Comment (aaltay) [inline sdks/python/apache_beam/examples/complete/juliaset/juliaset/juliaset.py]: This comment makes more sense next to where the pipeline is constructed. That is an opinion, feel free to ignore.

File: sdks/python/apache_beam/examples/complete/tfidf.py [hunk 3]
```diff
@@ -68,7 +68,8 @@ def expand(self, uri_to_content):
     # Create a collection of pairs mapping a URI to each of the words
     # in the document associated with that that URI.
 
-    def split_into_words((uri, line)):
+    def split_into_words(uri_line):
```
Comment (aaltay) [inline sdks/python/apache_beam/examples/complete/tfidf.py]: Is this because of https://www.python.org/dev/peps/pep-3113/? This is sad.

File: sdks/python/apache_beam/io/gcp/datastore/v1/datastoreio_test.py [hunk 4]
```diff
@@ -238,7 +241,7 @@ def fake_run_query(req):
       elif req == kind_stat_req:
         return kind_stat_resp
       else:
-        print kind_stat_req
+        print(kind_stat_req)
```
Comment (aaltay) [inline sdks/python/apache_beam/io/gcp/datastore/v1/datastoreio_test.py]: We have too many print statements. We should probably convert them to logs or remove them.

File: sdks/python/apache_beam/typehints/trivial_inference.py [hunk 5]
```diff
@@ -272,12 +279,12 @@ def infer_return_type_func(f, input_types, debug=False, depth=0):
     TypeInferenceError: if no type can be inferred.
   """
   if debug:
```
Comment (aaltay) [inline sdks/python/apache_beam/typehints/trivial_inference.py]: I think we should remove debug code. Unrelated to this PR, but worth filing a JIRA.

File: sdks/python/apache_beam/testing/util.py [hunk 6]
```diff
@@ -103,7 +103,7 @@ def expand(self, pcoll):
           | "ToVoidKey" >> Map(lambda v: (None, v)))
       _ = ((keyed_singleton, keyed_actual)
            | "Group" >> CoGroupByKey()
-           | "Unkey" >> Map(lambda (k, (_, actual_values)): actual_values)
+           | "Unkey" >> Map(lambda k___actual_values: k___actual_values[1][1])
```
Comment (aaltay) [inline sdks/python/apache_beam/testing/util.py]: why double _ here?

File: sdks/python/run_pylint.sh [hunk 7]
```diff
@@ -79,6 +80,16 @@ done
...
-pushd $MODULE
+pushd "$MODULE"
 isort -p apache_beam -w 120 -y -c -ot -cs -sl ${SKIP_PARAM}
 popd
+echo "Checking for files requiring stage 1 refactoring from futurize"
+futurize_results=$(futurize -j 8 --stage1 apache_beam 2>&1 |grep Refactored)
+futurize_filtered=$(echo "$futurize_results" |grep -v 'pb2\|typehints.py\|trivial_inference.py' || echo "")
+count=${#futurize_filtered}
+if [ "$count" != "0" ]; then
+  echo "Some of the changes require futurize stage 1 changes."
```
Comment (aaltay) [inline sdks/python/run_pylint.sh]: What does it mean? What action developer can take?

File: sdks/python/run_pylint.sh [hunk 8]
```diff
@@ -79,6 +80,16 @@ done
 for file in "${EXCLUDED_GENERATED_FILES[@]}"; do
   SKIP_PARAM="$SKIP_PARAM --skip $(basename $file)"
 done
-pushd $MODULE
+pushd "$MODULE"
 isort -p apache_beam -w 120 -y -c -ot -cs -sl ${SKIP_PARAM}
 popd
+echo "Checking for files requiring stage 1 refactoring from futurize"
+futurize_results=$(futurize -j 8 --stage1 apache_beam 2>&1 |grep Refactored)
+futurize_filtered=$(echo "$futurize_results" |grep -v 'pb2\|typehints.py\|trivial_inference.py' || echo "")
```
Comment (aaltay) [inline sdks/python/run_pylint.sh]: Could you separate the exclusion list?

File: sdks/python/apache_beam/typehints/typehints.py [hunk 9]
```diff
@@ -184,7 +185,17 @@ def bind_type_variables(type_constraint, bindings):
   return type_constraint
 
 
-class SequenceTypeConstraint(TypeConstraint):
+class IndexableTypeConstraint(TypeConstraint):
+  """An internal common base-class for all type constraints with indexing.
+  E.G. SequenceTypeConstraint + Tuple's of fixed size.
+  """
+
+  def _constraint_for_index(self, idx):
```
Comment (aaltay) [inline sdks/python/apache_beam/typehints/typehints.py]: What is this for?

**Comment by holdenk** [comment]
@aaltay So on the general case of tuple unpacking in lambdas: that's just gone in Python 3 and its not coming back.

In different places in the code the auto converter has taken different approaches to work around this.
If it's a function we can do multi-line you see things like "xxx_todo_renameme" (or you did before they were renamed but if you look at the commit from running futurize you can see it there) which is then followed by an explicit tuple unpack. In lambdas though that isn't really an option so we do the tuple indexing with compound names.

Some of the lambdas which were too long after auto convert were transformed into def functions and use tuple indexing. I can re-factor these ones to do the explicit tuple unpacking and avoid the indexing if you think it looks nicer.

File: sdks/python/apache_beam/typehints/typehints.py (hunk 9 above)
Comment (holdenk) [reply sdks/python/apache_beam/typehints/typehints.py]: Good point, I'll add a docstring. But if you look in simple_type_inference you can see where its used. Because of the insreased use of tuple indexing after conversion it's important that the type inference handle this correctly and the existing code was not.

File: sdks/python/run_pylint.sh (hunk 7 above)
Comment (holdenk) [reply sdks/python/run_pylint.sh]: So if this fails it will print out which files need changing. I'll add a further statement that running futurize can be used to see the proposed difference.

File: sdks/python/apache_beam/testing/util.py (hunk 6 above)
Comment (holdenk) [reply sdks/python/apache_beam/testing/util.py]: So one of the _s is from the original name (which indicated an ignored param). We could give it an explicit name or keep it with the _ to indicate unused.

File: sdks/python/apache_beam/typehints/trivial_inference.py (hunk 5 above)
Comment (holdenk) [reply sdks/python/apache_beam/typehints/trivial_inference.py]: I've found the debug code to be pretty helpful while working on fixing the changes in tirivial inference. I know its a bit weird but if it doesn't slow things down for folks I assume the trivial inference stuff is going to break again in the future so having a debug flag we can turn on can be useful. That being said we could just replace it with a comment that there used to be debugging code back at [X] and then people can go back and see it if they need it but w/e.

File: sdks/python/apache_beam/io/gcp/datastore/v1/datastoreio_test.py (hunk 4 above)
Comment (holdenk) [reply sdks/python/apache_beam/io/gcp/datastore/v1/datastoreio_test.py]: I agree, however I think that probably belongs in a follow up JIRA/PR. Do you agree?

File: sdks/python/apache_beam/examples/complete/tfidf.py (hunk 3 above)
Comment (holdenk) [reply sdks/python/apache_beam/examples/complete/tfidf.py]: Yup

File: sdks/python/apache_beam/examples/complete/game/leader_board.py (hunk 1 above)
Comment (holdenk) [reply sdks/python/apache_beam/examples/complete/game/leader_board.py]: Too long line after the unpacking change.

File: sdks/python/run_pylint.sh (hunk 8 above)
Comment (holdenk) [reply sdks/python/run_pylint.sh]: Sure, done :)

File: sdks/python/apache_beam/examples/complete/game/leader_board.py (hunk 1 above)
Comment (aaltay) [reply sdks/python/apache_beam/examples/complete/game/leader_board.py]: Why don't we break into two lines then?

**Comment by aaltay** [comment]
Thank you for clarifications @holdenk. My preference at least for public facing parts (apis, examples) to re-factor and not use the autoconverted version as-is. For other places I do not have a strong preference, although I think we can improve upon aa_bb_cc kind of naming. 

Some of these work can be follow up PRs as well.

Another question, from the commit history it is not clear what is an automated change, what is a manual change. Is there a way for me to tell which is which?

**Comment by holdenk** [comment]
Sure the automated changes are "Futurize stage 1 run", "re-order imports", and "Cleanup whitespace (autopep8 fix" -- the rest are by hand.

I'll dig through the api & examples directory a bit and see what I can do in the scope of this PR and what should maybe be pushed down the road.

File: sdks/python/apache_beam/examples/complete/game/leader_board.py (hunk 1 above)
Comment (holdenk) [reply sdks/python/apache_beam/examples/complete/game/leader_board.py]: multi-line lambdas http://www.artima.com/weblogs/viewpost.jsp?thread=147358

**Comment by aaltay** [comment]
R: @robertwb -- for continuing the review.

**Comment by holdenk** [comment]
Generally I'm trying to do minimal changes over top of the futurize tool (even then it's non-trivial) since we need to do a follow up with stage2 and then fix everything from that (which is going to on the bigger side). And then once more to fix straggler py3 issues. Keeping these PRs like this open for long running time has a non-zero cost since they touch a lot of files in trivial ways merge conflicts will pop up frequently.

**Comment by holdenk** [comment]
Gentle post weekend ping to @robertwb

**Comment by holdenk** [comment]
So it seems like I'm might be release managing for the Apache Spark project's next minor release which might take me away from having the time to work on beam until the release is out. If it's possible to get this reviewed before things get heavy on the Spark release would be awesome, but if not I understand.

**Comment by robertwb** [comment]
Yeah, I too have been swamped by some other stuff, and agreed that these are the kinds of changes that grow stale quickly. I'll try hunting around for another reviewer if I can't get to it right away (and of course will be happy if anyone else just jumps in :)).

**Comment by holdenk** [comment]
So surprisingly its not going stale as quickly as I thought (yay). I've got some time until Strata kicks off at which point I'll be back into release land.

File: sdks/python/apache_beam/runners/worker/operations.py [hunk 10]
```diff
@@ -434,7 +434,7 @@ def __init__(self, operation_name, spec, counter_factory, state_sampler):
     fn, args, kwargs = pickler.loads(self.spec.combine_fn)[:3]
     self.combine_fn = curry_combine_fn(fn, args, kwargs)
     if (getattr(fn.add_input, 'im_func', None)
-        is core.CombineFn.add_input.im_func):
+        is core.CombineFn.add_input.__func__):
```
[the remark this answers is not in the document]
Comment (holdenk) [reply sdks/python/apache_beam/runners/worker/operations.py]: Possibly not, but this is to only apply the "safe" automatic transformations -- there is still two follow ups.

**Comment by robertwb** [comment]
Thank you!
